File: typeScript/sorting/merge-sort.ts
const mergeSort = (arr: number[]): number[] => {
  if (arr.length <= 1) return arr;

  const middle = Math.floor(arr.length / 2);
  const left = mergeSort(arr.slice(0, middle));
  const right = mergeSort(arr.slice(middle));

  return merge(left, right);
};

const merge = (left: number[], right: number[]): number[] => {
  const result: number[] = [];

  while (left.length && right.length) {
    let toShift: number[] = [];
    if (left[0] <= right[0]) {
      toShift = left;
    } else {
      toShift = right;
    }

    if (toShift) {
      result.push(...toShift);
    }
  }

  while (left.length) {
    const leftShift = left.shift();
    if (leftShift) result.push(leftShift);
  }

  while (right.length) {
    const rightShift = right.shift();
    if (rightShift) result.push();
  }

  return result;
};

File: go/sorting/merge-sort.go
package sorting

func MergeSort(arr []int) []int {
	if len(arr) <= 1 {
		return arr
	}

	mid := len(arr) / 2
	left := MergeSort(arr[:mid])
	right := MergeSort(arr[mid:])

	return merge(left, right)
}

func merge(left, right []int) []int {
	var result []int

	for len(left) != 0 && len(right) != 0 {
		if left[0] <= right[0] {
			result = append(result, left[0])

			left = left[1:]
		} else {
			result = append(result, right[0])

			right = right[1:]
		}
	}

	for len(left) != 0 {
		result = append(result, left[0])

		left = left[1:]
	}

	for len(right) != 0 {
		result = append(result, right[0])

		right = right[1:]
	}

	return result
}
